From 566024d40b0dbcc7e4b1460f34b6fcd7b34e1370 Mon Sep 17 00:00:00 2001
From: Sidhanta Sahu <quic_sidhanta@quicinc.com>
Date: Thu, 6 Mar 2025 00:57:44 -0800
Subject: [PATCH 3/3] QSDK:CP : wifi: ath12k: Fix to get sla_mask for partner
 peer links

Currently, sla_mask is only provided for primary links and not passed
for partner links. The sla_mask is obtained from the ath12k_peer's sawf
peer context, which is maintained only for primary link peers and not
updated for partner links.

Fix this by obtaining the primary ath12k_peer from the partner's
ath12k_peer object to fetch the sawf peer context.

Patch-dependency: ath12k/983-QSDK-CP-wifi-ath12k-Populate-RSSI-SNR-sla-mask-in-AT.patch
Patch-work: None

Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/sawf.c               | 9 +++++++--
 drivers/net/wireless/ath/ath12k/sawf.h               | 2 +-
 drivers/net/wireless/ath/ath12k/telemetry_agent_if.c | 7 +++++--
 3 files changed, 13 insertions(+), 5 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/sawf.c
+++ b/drivers/net/wireless/ath/ath12k/sawf.c
@@ -148,18 +148,23 @@ ath12k_get_enabled_sawf_params_mask(stru
 	return enabled_sawf_params;
 }
 
-void ath12k_get_peer_sla_config(struct ath12k_peer *peer, u16 *sla_mask)
+void ath12k_get_peer_sla_config(const u8 *sta_addr, u16 *sla_mask)
 {
 	struct ath12k_sawf_peer_ctx *peer_ctx = NULL;
 	struct ath12k_sawf_ctx *sawf_ctx;
+	struct ath12k_peer *pri_peer;
 	int tid, q_id;
 	u8 svc_id;
 
+	pri_peer = ath12k_peer_find_primary_by_sta_addr(sta_addr);
+	if (!pri_peer)
+		return;
+
 	sawf_ctx = ath12k_get_sawf_context();
 	if (!sawf_ctx)
 		return;
 
-	peer_ctx = &peer->sawf_ctx_peer;
+	peer_ctx = &pri_peer->sawf_ctx_peer;
 
 	for (tid = 0; tid < ATH12K_SAWF_MAX_TID_SUPPORT; tid++) {
 		for (q_id = 0; q_id < MAX_Q_PER_TID; q_id++) {
--- a/drivers/net/wireless/ath/ath12k/sawf.h
+++ b/drivers/net/wireless/ath/ath12k/sawf.h
@@ -908,7 +908,7 @@ void ath12k_sawf_stats_update(struct ath
 			      struct hal_tx_status *ts,
 			      struct ath12k_peer *peer, ktime_t timestamp);
 
-void ath12k_get_peer_sla_config(struct ath12k_peer *peer, u16 *sla_mask);
+void ath12k_get_peer_sla_config(const u8 *sta_addr, u16 *sla_mask);
 
 void ath12k_sawf_tx_enqueue_peer_stats(struct ath12k_base *ab,
 				       struct ath12k_peer *peer,
--- a/drivers/net/wireless/ath/ath12k/telemetry_agent_if.c
+++ b/drivers/net/wireless/ath/ath12k/telemetry_agent_if.c
@@ -251,6 +251,7 @@ int ath12k_get_peer_stats(void *obj, str
 	}
 
 	rcu_read_lock();
+
 	peer = ath12k_peer_find_by_peer_list_index(pdev->ar, peer_db->peer_id);
 	if (!peer) {
 		rcu_read_unlock();
@@ -282,8 +283,8 @@ int ath12k_get_peer_stats(void *obj, str
 	}
 
 	stats->rssi = ath12k_calculate_link_rssi(peer);
-	if (peer->primary_link)
-		ath12k_get_peer_sla_config(peer, &stats->sla_mask);
+	ath12k_get_peer_sla_config(peer->mlo ? stats->peer_mld_mac : stats->peer_link_mac,
+				   &stats->sla_mask);
 	rcu_read_unlock();
 	for (ac = 0; ac < ATH12K_DP_WLAN_MAX_AC; ac++) {
 		stats->airtime_consumption[ac] =
